reports/tasks/msisdn_message_report.py

```diff
@@ -13,63 +16,200 @@ class GenerateMSISDNMessageReport(BaseTask):
+        )
+
+        data = self.retrieve_identity_info(is_client, msisdns)
+
+        data = self.retrieve_registration_info(is_client, data)
+
+        (data, list_length) = self.retrieve_messages(ms_client, data,
+                                                     start_date, end_date)
+
+        spreadsheet = self.populate_spreadsheet(msisdns, data, list_length)
+
         output_file = generate_random_filename()
         spreadsheet.save(output_file)
 
         task_status.status = ReportTaskStatus.SENDING
         task_status.file_size = getsize(output_file)
         task_status.save()
 
-        # TODO: Use SendEmail task to send the email
-
-    def create_spreadsheet(self, msisdns):
+        if email_recipients:
+            file_name = 'msisdn-report-%s-to-%s.xlsx' % (
+                start_date.strftime('%Y-%m-%d'),
+                end_date.strftime('%Y-%m-%d'))
+            SendEmail.apply_async(kwargs={
+                'subject': email_subject,
+                'file_name': file_name,
+                'file_location': output_file,
+                'sender': email_sender,
+                'recipients': email_recipients,
+                'task_status_id': task_status_id})
+
+    def retrieve_identity_info(self, is_client, msisdns):
         logger = self.get_logger()
-        workbook = ExportWorkbook()
-        sheet = workbook.add_sheet('Data for study cohort', 0)
-
-        identity_store = IdentityStoreApiClient(
-            settings.IDENTITY_STORE_TOKEN,
-            settings.IDENTITY_STORE_URL,
-        )
-
-        sheet.set_header([
-            'Phone number',
-            'Date registered',
-            'Facility',
-            'Pregnancy week',
-            'Message type',
-            'Message 1: content',
-            'Message 1: date sent',
-            'Message 1: status',
-        ])
 
+        data = {}
         for msisdn in msisdns:
-            response = identity_store.get_identity_by_address('msisdn', msisdn)
+            response = is_client.get_identity_by_address('msisdn', msisdn)
             results = list(response['results'])
 
             if len(results) < 1:
                 logger.info(
                     'No results from identity store for {0}'.format(msisdn))
+                data[msisdn] = {}
                 continue
 
             if len(results) > 1:
                 logger.info(
                     'Multiple results from identity store for {0}'
                     .format(msisdn))
+                data[msisdn] = {}
                 continue
 
-            identity = results[0]
+            data[msisdn] = {'id': results[0]['id'],
+                            'created_at': results[0]['created_at']}
 
-            sheet.add_row({
-                'Phone number': msisdn,
-                'Date registered': identity['created_at'],
+        return data
+
+    def retrieve_registration_info(self, is_client, data):
+        logger = self.get_logger()
+
+        for msisdn in data:
```

Comment:
Since each value in the data is a dictionary, we can probably do a `for msisdn, datum in data.items()`, and then when updating the values below, instead of doing the `data[msisdn]['key'] = 'value'`, we can just do `datum['key'] = 'value'`